src/outputs/rcp.c

```diff
@@ -0,0 +1,1484 @@
+  int ret;
+
+  DPRINTF(E_DBG, L_RCP, "device volume set %ld '%s' state=%d vol=%d\n", device->id, device->name, s->state, device->volume);
+
+  if (!s || (s && s->state != RCP_STATE_STREAMING))
+    return 0;
+
+  s->callback_id = callback_id;
+
+
+  ret = snprintf(cmd, sizeof(cmd), "%d", device->volume);
+  if (ret < 0) {
+    return 0;
+  }
+
+  rcp_send(s, RCP_STATE_VOL_SET, cmd);
+
+  rcp_status(s);
+
+  return 1;
+}
+
+static void
+rcp_device_cb_set(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s = device->session;
+
+  DPRINTF(E_DBG, L_RCP, "device cb %ld '%s'  callback %d\n", device->id, device->name, callback_id);
+
+  s->callback_id = callback_id;
+}
+
+static void
+rcp_device_add(const char *device_name, const char *address)
+{
+  struct output_device *device;
+  uint64_t id;
+  int ret;
+
+  /* possible if this was mdns/avahi that there is a timeout/cached entry is no
+   * longer resolvable;  ie avahi caches a roku name, the roku is pwoered down,
+   * cache entry still exists but now no address
+   */ 
+  if (!address)
+    {
+      DPRINTF(E_DBG, L_RCP, "Attempt to add '%s' without address (possible mdns timeout)\n", device_name ? device_name : "<no name>");
+      return;
+    }
+
+  id = djb_hash(address, strlen(address));
+  if (!id)
+    {
+      DPRINTF(E_LOG, L_RCP, "Could not hash RCP adddress name for id (%s)\n", address);
+      return;
+    }
+
+  CHECK_NULL(L_RCP, device = calloc(1, sizeof(struct output_device)));
+
+  // 3.0.44 being the last public firmware of the Roku Soundbridge devices...
+  device->id = id + 3044;
+  device->name = strdup(device_name);
+  device->type = OUTPUT_TYPE_RCP;
+  device->type_name = outputs_name(device->type);
+  device->has_video = 0;
+  device->v4_address = strdup(address);
+  DPRINTF(E_INFO, L_RCP, "Adding RCP output device '%s' at '%s'\n", device_name, address);
+
+  ret = player_device_add(device);
+  if (ret < 0)
+    outputs_device_free(device);
+}
+
+static void
+rcp_mdns_device_cb(const char *name, const char *type, const char *domain, const char *hostname, int family, const char *address, int port, struct keyval *txt)
+{
+  bool exclude;
+
+  /* $ avahi-browse -vrt  _roku-rcp._tcp
+	Server version: avahi 0.7; Host name: foo.local
+	E Ifce Prot Name                       Type                 Domain
+	+  eth0 IPv4 SoundBridge              _roku-rcp._tcp       local
+	=  eth0 IPv4 SoundBridge              _roku-rcp._tcp       local
+	   hostname = [SoundBridge.local]
+	   address = [192.168.0.3]
+	   port = [5555]
+	   txt = []
+	: Cache exhausted
+	: All for now
+   */
+
+  exclude = cfg_getbool(cfg_gettsec(cfg, "rcp", name), "exclude");
+  DPRINTF(E_DBG, L_RCP, "Event for %sRCP/SoundBridge device '%s' (address %s, port %d)\n", exclude ? "excluded " : "", name, address, port);
+  
+  if (exclude)
+    DPRINTF(E_INFO, L_RCP, "Excluding discovered RCP/SoundBridge device '%s' at %s\n", name, address);
+  else
+    rcp_device_add(name, address);
```

Comment:
What if the device is powered off/disappears (the port is -1)? Should be removed in that case.

Reply:
when the remote device disappears, i;ve seen avahi announce a `NULL` address for a previously seen device which we take care of in `rcp_device_add`